Userland/Libraries/LibCore/Promise.h

```diff
@@ -1,76 +1,109 @@
 #include <LibCore/Object.h>
 
 namespace Core {
 
-template<typename Result>
+template<typename Result, typename TError>
 class Promise : public Object {
     C_OBJECT(Promise);
 
 public:
-    Function<ErrorOr<void>(Result&)> on_resolved;
+    using ErrorType = TError;
 
-    ErrorOr<void> resolve(Result&& result)
+    Function<ErrorOr<void>(Result&)> on_resolution;
+    Function<void(ErrorType&)> on_rejection;
+
+    void resolve(Result&& result)
     {
-        m_pending_or_error = move(result);
+        m_result_or_rejection = move(result);
 
-        if (on_resolved)
-            return on_resolved(m_pending_or_error->value());
-        return {};
+        if (on_resolution) {
+            auto handler_result = on_resolution(m_result_or_rejection->value());
+            possibly_handle_rejection(handler_result);
+        }
     }
 
-    void cancel(Error error)
+    void reject(ErrorType&& error)
     {
-        m_pending_or_error = move(error);
+        m_result_or_rejection = move(error);
+        possibly_handle_rejection(*m_result_or_rejection);
     }
 
-    bool is_canceled()
+    bool is_rejected()
     {
-        return m_pending_or_error.has_value() && m_pending_or_error->is_error();
+        return m_result_or_rejection.has_value() && m_result_or_rejection->is_error();
     }
 
     bool is_resolved() const
     {
-        return m_pending_or_error.has_value() && !m_pending_or_error->is_error();
+        return m_result_or_rejection.has_value() && !m_result_or_rejection->is_error();
     }
 
-    ErrorOr<Result> await()
+    ErrorOr<Result, ErrorType>& await()
     {
-        while (!m_pending_or_error.has_value())
+        while (!m_result_or_rejection.has_value())
             Core::EventLoop::current().pump();
 
-        return m_pending_or_error.release_value();
+        return *m_result_or_rejection;
     }
 
     // Converts a Promise<A> to a Promise<B> using a function func: A -> B
     template<typename T>
-    RefPtr<Promise<T>> map(T func(Result&))
+    RefPtr<Promise<T>> map(Function<T(Result&)> func)
     {
         RefPtr<Promise<T>> new_promise = Promise<T>::construct();
-        on_resolved = [new_promise, func](Result& result) {
-            auto t = func(result);
-            return new_promise->resolve(move(t));
+        on_resolution = [new_promise, func = move(func)](Result& result) -> ErrorOr<void> {
+            new_promise->resolve(func(result));
+            return {};
+        };
+        on_rejection = [new_promise](ErrorType& error) {
+            new_promise->reject(move(error));
         };
         return new_promise;
     }
 
+    template<typename F>
+    requires(IsCallableWithArguments<F, ErrorOr<void>, Result&>)
+    Promise& when_resolved(F handler)
+    {
+        on_resolution = move(handler);
+        return *this;
+    }
+
+    template<typename F>
+    requires(IsCallableWithArguments<F, void, ErrorType&>)
+    Promise& when_rejected(F handler)
+    {
+        on_rejection = move(handler);
```

Comment:
If the promise is running on another thread, `m_result_or_rejection` may already contain something for us to process here. Would it be possible to guarantee that `on_rejection()` is called in this function when that's the case?

Reply:
hmm...sure, can do a simple check here, mutex though...you might have to put the whole promise behind one

Reply:
I was thinking perhaps we could take the `Optional<ErrorOr<Result, ErrorType>>` out and instead store something equivalent to an atomic-protected `Variant<Empty, Result, ErrorType>`:

```cpp
enum class ContentsType {
    Empty,
    HasResult,
    HasError,
};

Atomic<Contents> m_contents_type;
union {
    Result result;
    ErrorType error;
} m_contents;
```

We could potentially even store raw Contents and use the atomic builtin wrappers to check the contents based on a template parameter.

I'm not expert on atomics, though, so I can't say I know for sure this works as well as a mutex.

Reply:
:thinking:
we could make a `Core::ThreadSafePromise` that incorporates a mutex, but what's wrong with `MutexProtected<Core::Promise<foo>>`?